### PR TITLE
BE-797 Sonarqube Integration

### PR DESCRIPTION
**Pull request by pec1985**
Sonarqube integration.

Depends on https://github.com/pinpt/httpclient/pull/1

### CODE REVIEW

File: integrations/sonarqube/api/http.go [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+	type metricsResponse struct {
+		Measures []*Measure `json:"measures"`
+	}
+
+	metricKeys := metrics
+	if fromDate != "" {
+		fromDate = "&from=" + fromDate
+	}
+
+	ur := "/measures/search_history?p=1&ps=500&component=" + componentID + "&metrics=" + metricKeys + fromDate
```
Comment (pec1985) [inline integrations/sonarqube/api/http.go]: The API allows up to 1000 
https://sonarcloud.io/web_api/api/measures/search_history

File: integrations/sonarqube/api/http.go [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+					r = append(r, h)
+				}
+			}
+			m.History = r
+		}
+	}
+
+	if err != nil {
+		fmt.Println(err)
+		return nil, nil
```
Comment (pec1985) [inline integrations/sonarqube/api/http.go]: Remove `fmt.Println()` and return `err`

File: integrations/sonarqube/api/http_test.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+	assert.NotEmpty(projects)
+
+	for _, p := range projects {
+		assert.NotEmpty(p.ID)
+		assert.NotEmpty(p.Name)
+		assert.NotEmpty(p.Key)
+	}
+}
+
+func TestFetchMetrics(t *testing.T) {
```
Comment (pec1985) [inline integrations/sonarqube/api/http_test.go]: This test is failing

File: integrations/sonarqube/export.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+)
+
+func export(ctx context.Context, logger log.Logger, outdir, metrics string, a api.SonarqubeAPI, config *integration.ExportConfig, progress integration.ExportProgress, metadata map[string]interface{}) error {
+
+	start := time.Now()
+	projects, err := fetchAndExportProjects(ctx, logger, outdir, a)
+	if err != nil {
+		return err
+	}
+	_, err = fetchAndExportMetrics(ctx, logger, outdir, a, projects, "complexity,code_smells,new_code_smells,sqale_rating,reliability_rating,security_rating,coverage,new_coverage,test_success_density,new_technical_debt")
```
Comment (pec1985) [inline integrations/sonarqube/export.go]: the _metrics_ params is hard-coded

File: pkg/ids/ids.go [hunk 5]
```diff
@@ -66,3 +69,28 @@ func MaskValues(values ...string) []string {
...
+			o2 = o.(time.Time).UTC().Format(time.RFC3339Nano)
+		case string, int64:
+			o2 = o
+		default:
+			panic(fmt.Errorf("type not supported as id %T", o))
+		}
+		objs2 = append(objs2, o2)
+	}
+	return HashValues(objs2...)
+}
```
Comment (pec1985) [inline pkg/ids/ids.go]: We should remove `HashValues` and `IDFromValues` and use `hash.Values()` instead
https://github.com/pinpt/agent/commit/d9ced02d34dcd1db7d0c72fd7bfa756b459c463a#r30207729

File: integrations/sonarqube/api/http.go [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+		}
+		for _, m := range mt {
+			for _, h := range m.History {
+				d, e := time.Parse("2006-01-02T15:04:05-0700", h.Date)
+				date := int64(0)
+				if e != nil {
+					return nil, err
+				}
+				date = d.Unix() * 1000
+				all = append(all, &Metric{
```
Comment (pec1985) [inline integrations/sonarqube/api/http.go]: we should use datetime package from go-common where possible for both the parsing and the conversion from unix to int64. https://github.com/pinpt/agent/commit/d9ced02d34dcd1db7d0c72fd7bfa756b459c463a#r30207698

File: integrations/sonarqube/api/http.go [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+					ComponentID:  proj.ID,
+					ComponentKey: proj.Key,
+					ID:           fmt.Sprintf("%v.%v.%v", proj.ID, date, m.Metric),
+					Date:         date,
+					Metric:       m.Metric,
+					Value:        h.Value,
```
Comment (pec1985) [inline integrations/sonarqube/api/http.go]: all our IDs need to be hashed using hash.Value from go-common
https://github.com/pinpt/agent/commit/d9ced02d34dcd1db7d0c72fd7bfa756b459c463a#r30207702

File: integrations/sonarqube/api/http.go [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+		for _, m := range mt {
+			for _, h := range m.History {
+				d, e := time.Parse("2006-01-02T15:04:05-0700", h.Date)
+				date := int64(0)
+				if e != nil {
+					return nil, err
+				}
+				date = d.Unix() * 1000
+				all = append(all, &Metric{
+					ComponentID:  proj.ID,
```
Comment (pec1985) [inline integrations/sonarqube/api/http.go]: all our IDs need to be hashed using hash.Value from go-common
https://github.com/pinpt/agent/commit/d9ced02d34dcd1db7d0c72fd7bfa756b459c463a#r30207702

File: integrations/sonarqube/sonarqube.go [hunk 9]
```diff
@@ -73,7 +73,8 @@ func (i *Integration) newAPI(ctx context.Context, logger log.Logger) (context.Co
 	apikey, _ := integration.GetValueString(i, "apikey")
 	ctx = api.ContextWithAuthorizationToken(ctx, apikey)
 	ctx = api.ContextWithSonarqubeURLPrefix(ctx, url)
-	a, err := api.NewHTTPAPI(ctx, logger)
+	s := integration.NewStorage(i, nil)
+	a, err := api.NewHTTPAPI(ctx, logger, s)
```
Comment (jhaynie) [inline integrations/sonarqube/sonarqube.go]: we're passing nil for the config ... is this correct?

**Pull request merged**